syntax = "proto3";

package simulation;

import "google/protobuf/timestamp.proto";

message Simulation {
    message SimulationDetails {
        message Range {
            optional double min = 1;
            optional double max = 2;
            optional uint32 bins = 3;
        }

        optional Range radialExtent = 1;
        optional Range heightExtent = 2;

        optional google.protobuf.Timestamp timeStamp = 3; // check

        message Layer {
            optional double n = 1;
            optional double height = 2;
            optional double mu_a = 3;
            optional double mu_s = 4;
            optional double g = 5;
        }

        repeated Layer layers = 4;

        optional uint64 photonsLaunched = 5;
    }

    optional SimulationDetails details = 1;
    repeated double reflectance = 2 [packed=true];
    repeated double absorption = 3 [packed=true]; // nb: 1D array to be interpreted as 2D
}

message ParameterSimulation {
    repeated Simulation simulations = 1;

    message Layer {
        optional double n = 1;
        optional double height = 2;
        optional double g = 3;

        // constant layers define these
        optional double mu_a = 4;
        optional double mu_s = 5;

        // wavelength dependent scattering
        optional double a = 6;
        optional double b = 7;

        // wavelength dependent absorption
        optional double cBlood = 8;
        optional double fOxy = 9;
        optional double fWater = 10;
        optional double fLipid = 11;
    }

    repeated Layer layers = 2;
    repeated double wavelengths = 3;
}